src/dom/cached-element-vector.ts

```diff
@@ -0,0 +1,144 @@
+/**
+ * CachedElementVector is used to optimize access to commonly checked values.
+ * Abstracts the caching of values such as window dimensions or scroll positions
+ * that can be stored as vectors.
+ */
+import { DynamicDefaultMap } from '../map/dynamic-default';
+import { MultiValueDynamicDefaultMap } from '../map/multi-value-dynamic-default';
+import { MultiDimensionalVector } from '../mathf/geometry/multi-dimensional-vector';
+import { Raf } from '..';
+
+// Number of past values to cache
+const VALUE_LIMIT: number = 2;
+
+type InnerCache = MultiValueDynamicDefaultMap<any, any>;
+
+/**
+ * All instances of CachedElementVector will share a single cache to facilitate
+ * abstraction of the cache checking/clearing.
+ *
+ * Also helps ensure that only singletons are being created by checking for
+ * existing instances in this cache.
+ */
+const caches: DynamicDefaultMap<any, InnerCache> =
+    DynamicDefaultMap.usingFunction<any, any>(
```

Comment:
this code is pretty hard for me to read, maybe because there's so much generics here, and maybe the api for the "default map" could be improved.

i wonder if you can move the generics handling to where the map values are being fetched instead, which would simplify things a little and make it generally easier to read. or maybe use a regular constructor so that the generics type setting isn't happening so much in the function call.

something like:

```javascript
class DefaultMap extends Map {
  getOrSet(key: string, value: any | () => any): any { ... }
}

const map = new DefaultMap();
const value = map.getOrSet('key', () => { return someDefaultValue }) as SomeType;
```

i think with something like this, i'd be easier for me to understand, like "ok this is a map, and if a value doesn't exist, it sets it."

with the example above i used `any` but generics could/should be used, i just didn't know the syntax off the top of my head. i'm guessing the user call would be something like:

```javascript
const map: DefaultMap<string, SomeType> = new DefaultMap();
const value = map.getOrSet('key', () => { return someDefaultValue });  // returns a `SomeType`
```

Reply:
This would go against the intent of the default map which is to avoid duplicated conditional get calls by assuming them as the default.

I'll remove this base class and re-implement its functions in its child classes as needed since this follows the same pattern as MultiDimensionalVector does.

Reply:
i could be wrong, but i think if you wrap the call in a fn, it won't get evaluated until the callback is called.

Reply:
It wasn't a performance concern as much as it was a developer ergonomics concern. I found I was doing a lot of boilerplate, particularly with caching, that could be eliminated with this class.

Part of the reason for writing this class was to avoid repeating `map.get(foo) || (foo) => someThingWithFoo(foo)`, or having to add private methods to each class to the same effect.